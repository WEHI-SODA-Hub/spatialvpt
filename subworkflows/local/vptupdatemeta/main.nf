include { VIZGENPOSTPROCESSING_PARTITIONTRANSCRIPTS    } from '../../../modules/local/vizgenpostprocessing/partitiontranscripts/main'
include { VIZGENPOSTPROCESSING_DERIVEENTITYMETADATA    } from '../../../modules/local/vizgenpostprocessing/deriveentitymetadata/main'
include { VIZGENPOSTPROCESSING_UPDATEVZG               } from '../../../modules/local/vizgenpostprocessing/updatevzg/main'

workflow VPTUPDATEMETA {

    take:
    micron_space
    detected_txs
    update_vzg
    input_vzg

    main:

    ch_versions = Channel.empty()

    //
    // MODULE: Run vpt derive-entity-metadata
    //
    VIZGENPOSTPROCESSING_DERIVEENTITYMETADATA(
        micron_space
    )

    //
    // MODULE: Run vpt partition-transcripts
    //
    VIZGENPOSTPROCESSING_PARTITIONTRANSCRIPTS(
        micron_space,
        detected_txs
    )

    // Output channels
    ch_entity_metadata =
        VIZGENPOSTPROCESSING_DERIVEENTITYMETADATA.out.entity_metadata

    ch_entity_by_gene =
        VIZGENPOSTPROCESSING_PARTITIONTRANSCRIPTS.out.transcripts

    ch_vzg = Channel.empty()
    if (update_vzg.value) {
        //
        // MODULE: Run vpt update-vzg
        //
        VIZGENPOSTPROCESSING_UPDATEVZG(
            micron_space.map { meta, _micron_space -> meta },
            input_vzg,
            micron_space,
            ch_entity_by_gene,
            ch_entity_metadata
        )

        ch_vzg = VIZGENPOSTPROCESSING_UPDATEVZG.out.vzg_file
    }

    // get transcripts channel for downstream output
    micron_space
    .map { meta, _micron_space -> meta }.concat(detected_txs)
    .set{ ch_transcripts }

    emit:
    metadata       = ch_entity_metadata
    transcripts    = ch_transcripts
    vzg            = ch_vzg

    versions = ch_versions                     // channel: [ versions.yml ]
}
