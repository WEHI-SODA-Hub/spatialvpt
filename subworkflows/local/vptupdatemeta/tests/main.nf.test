nextflow_workflow {

    name "Test Subworkflow VPTUPDATEMETA"
    script "../main.nf"
    workflow "VPTUPDATEMETA"

    tag "subworkflows"
    tag "subworkflows/vptupdatemeta"

    tag "vizgenpostprocessing/deriveentitymetadata"
    tag "vizgenpostprocessing/partitiontranscripts"
    tag "vizgenpostprocessing/updatevzg"

    test("merscope - parquet, csv, vzg") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test' ],
                    file('${projectDir}/tests/data/cellpose2_micron_space.parquet')
                ])
                input[1] = file('${projectDir}/tests/data/minimal_data/detected_transcripts.csv')
                input[2] = Channel.value(false) // Do not update vzg
                input[3] = file("dummy.vzg", checkIfExists: false) // Dummy vzg file
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}
