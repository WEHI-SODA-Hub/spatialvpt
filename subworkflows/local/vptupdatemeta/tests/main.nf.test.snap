{
    "merscope - parquet, csv, vzg": {
        "content": [
            {
                "0": [
                    "cell_metadata_resegmented.csv:md5,cb734bdf1dbcbd65015880eb4fafc64b"
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "detected_transcripts.csv:md5,65ecb4084dc540dba17b5b1429924da1"
                    ]
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "metadata": [
                    "cell_metadata_resegmented.csv:md5,cb734bdf1dbcbd65015880eb4fafc64b"
                ],
                "transcripts": [
                    [
                        {
                            "id": "test"
                        },
                        "detected_transcripts.csv:md5,65ecb4084dc540dba17b5b1429924da1"
                    ]
                ],
                "versions": [
                    
                ],
                "vzg": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-08-26T12:20:40.19847494"
    }
}